chemprop/cli/train.py

```diff
@@ -652,6 +615,58 @@ def main(args):
         args.max_lr,
         args.final_lr,
     )
+
+
+def main(args):
+    format_kwargs = dict(
+        no_header_row=args.no_header_row,
+        smiles_cols=args.smiles_columns,
+        rxn_cols=args.reaction_columns,
+        target_cols=args.target_columns,
+        ignore_cols=args.ignore_columns,
+        weight_col=args.weight_column,
+        bounded=args.loss_function is not None and "bounded" in args.loss_function,
+    )
+    featurization_kwargs = dict(
+        features_generators=args.features_generators, keep_h=args.keep_h, add_h=args.add_h
+    )
+
+    all_data = build_data_from_files(
+        args.data_path,
+        p_features=args.features_path,
+        p_atom_feats=args.atom_features_path,
+        p_bond_feats=args.bond_features_path,
+        p_atom_descs=args.atom_descriptors_path,
+        **format_kwargs,
+        **featurization_kwargs,
+    )
+
+    multicomponent = len(all_data) > 1
+    if not multicomponent:
+        all_data = all_data[0]
+
+    train_data, val_data, test_data = build_splits(
+        args, all_data, format_kwargs, featurization_kwargs
+    )
+    train_dset, val_dset, test_dset = build_datasets(
+        args, multicomponent, train_data, val_data, test_data
```

Comment:
Why does `build_datasets()` take `multicomponent` while `build_splits()` does not? `build_datasets()` could calculate it itself like `build_splits()` does. `multicomponent = len(train_data) > 1`. Or `build_splits()` could take `multicomponent` as an argument. I think consistency enhances code readability here.

Reply:
I see now that `multicomponent = len(train_data) > 1` will not work because if it is multicomponent `train_data` will be a single list of MoleculeDatapoint's. Instead we could check if the first element in train_data is a list. Open to other ideas.

Reply:
Also `build_splits()` did its `multicomponent = len(all_data) > 1` check _after_ `all_data = all_data[0]` in `main()` which meant `build_splits()` would always see `len(all_data) > 1` as true. We could pass `multicomponent` into `build_splits()` to fix that. I instead just moved the `all_data` creation into `build_splits()` because that made more sense to me. I'm open to changing it back, and I didn't change the name of `build_splits()`.

Reply:
I really hate type-checking as a logical pattern in Python. It's almost completely against the design pattern of the language is extremely brittle. Consider the possibility where someone changes the return type of `build_splits` to be a `tuple[tuple[Dataset, ...], ...] | tuple[Dataset, ...]`, i.e., changes the inner `list` to a `tuple`. That would break this type check, so I decided it was better to be explicit in the code and pass in the `multicomponent` argument.